src/main/java/com/microsoft/azure/vmagent/AzureVMManagementServiceDelegate.java

```diff
@@ -1692,36 +1694,54 @@ public boolean isVMAliveOrHealthy(AzureVMAgent agent) throws AzureCloudException
      * @return Total VM count
      */
     public int getVirtualMachineCount(String cloudName, String resourceGroupName) {
+    	final Map<String, Integer> countsByTemplate = getVirtualMachineCountsByTemplate(cloudName, resourceGroupName);
+        int count = 0;
+        for (Integer countForTemplate : countsByTemplate.values() ) {
+        	count += countForTemplate.intValue();
+        }
+        return count;
+    }
+
+    /**
+     * Retrieves count of virtual machine in a azure subscription indexed by template name.
+     *
+     * @return {@link Map} of {@link String} template name to {@link Integer} count of VMs from that template.
+     * @throws ManagementException if we get an Azure error that isn't 404.
+     */
+    @Restricted(NoExternalUse.class)
+    Map<String, Integer> getVirtualMachineCountsByTemplate(final String cloudName, final String resourceGroupName) {
```

Comment:
I've not added unit-tests for this method.
It'll be exercised by the existing unit-tests for the original getVirtualMachineCount method, but there's nothing specific for this one at present.
...and I'd need a few hints as to how best to do that, as it'll need to "fake" a call to Azure (which the existing unit tests manage, but I don't fully understand how).